In [1]:
import pandas as pd
import numpy as np

print("import successful")

import successful


Importing Libraries 

In [2]:
import pandas as pd
import numpy as np
from matplotlib import pyplot as plt
import seaborn as sns
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import classification_report, confusion_matrix, precision_score, recall_score, f1_score
from sklearn.model_selection import cross_val_score
from sklearn.model_selection import GridSearchCV
from sklearn.preprocessing import StandardScaler, OneHotEncoder
from sklearn.compose import ColumnTransformer
from sklearn.pipeline import Pipeline
from sklearn.impute import SimpleImputer
from sklearn.ensemble import RandomForestClassifier
from xgboost import XGBClassifier
from sklearn.ensemble import GradientBoostingClassifier
from sklearn.model_selection import RandomizedSearchCV

Matplotlib is building the font cache; this may take a moment.


In [3]:
# Loading the dataset
customer_transaction_data = pd.read_csv(r'/Users/Dell/Frudulent-transaction-detection-for-Finlora-Company/Finlora_Dataset/FinLora_Customer_Transaction_Dataset.csv')
customer_transaction_data.head()

,transaction_id,customer_id,timestamp,home_country,source_currency,dest_currency,channel,amount_src,amount_usd,fee,...,ip_risk_score,kyc_tier,account_age_days,device_trust_score,chargeback_history_count,risk_score_internal,txn_velocity_1h,txn_velocity_24h,corridor_risk,is_fraud
0,fee8542d-8ee6-4b0d-9671-c294dd08ed26,402cccc9-28de-45b3-9af7-cc5302aa1f93,2022-10-03 18:40:59.468549+00:00,US,USD,CAD,ATM,278.19,278.19,4.25,...,0.123,standard,263,0.522,0,0.223,0,0,0.0,0
1,bfdb9fc1-27fe-4a85-b043-4d813d679259,67c2c6b3-ef0a-4777-a3f1-c84a851bb6ad,2022-10-03 20:39:38.468549+00:00,CA,CAD,MXN,web,208.51,154.29,4.24,...,0.569,standard,947,0.475,0,0.268,0,1,0.0,0
2,fc855034-3ea5-4993-9afa-b511d93fe5e8,6d0d9b27-fa26-45f8-93b1-2df29d182d9c,2022-10-03 23:02:43.468549+00:00,US,USD,CNY,mobile,160.33,160.33,2.70,...,0.437,enhanced,367,0.939,0,0.176,0,0,0.0,0
3,2cf8c08e-42ec-444d-a755-34b9a2a0a4ca,7bd5200c-5d19-44f0-9afe-8b339a05366b,2022-10-04 01:08:53.468549+00:00,US,USD,EUR,mobile,59.41,59.41,2.22,...,0.594,standard,147,0.551,0,0.391,0,0,0.0,0
4,d907a74d-b426-438d-97eb-dbe911aca91c,70a93d26-8e3a-4179-900c-a4a7a74d08e5,2022-10-04 09:35:03.468549+00:00,US,USD,INR,mobile,200.96,200.96,3.61,...,0.121,enhanced,257,0.894,0,0.257,0,0,0.0,0


In [4]:
customer_transaction_data.shape

(11400, 26)

In [6]:
#checking for all the columns in the dataset
customer_transaction_data.columns

Index(['transaction_id', 'customer_id', 'timestamp', 'home_country',
       'source_currency', 'dest_currency', 'channel', 'amount_src',
       'amount_usd', 'fee', 'exchange_rate_src_to_dest', 'device_id',
       'new_device', 'ip_address', 'ip_country', 'location_mismatch',
       'ip_risk_score', 'kyc_tier', 'account_age_days', 'device_trust_score',
       'chargeback_history_count', 'risk_score_internal', 'txn_velocity_1h',
       'txn_velocity_24h', 'corridor_risk', 'is_fraud'],
      dtype='object')

In [7]:
#checking for missing values
customer_transaction_data.isnull().sum()

transaction_id                 0
customer_id                    0
timestamp                     29
home_country                   0
source_currency                0
dest_currency                  0
channel                        0
amount_src                     0
amount_usd                   305
fee                          295
exchange_rate_src_to_dest      0
device_id                      0
new_device                     0
ip_address                   305
ip_country                   301
location_mismatch              0
ip_risk_score                  0
kyc_tier                     300
account_age_days               0
device_trust_score           295
chargeback_history_count       0
risk_score_internal            0
txn_velocity_1h                0
txn_velocity_24h               0
corridor_risk                  0
is_fraud                       0
dtype: int64

DATA DICTIONARY

1. transaction_id

Data Type: Object
Description: Unique identifier assigned to each transaction.
Example: TXN_000001

2. customer_id

Data Type: Object
Description: Unique identifier assigned to each customer associated with the transaction.
Example: CUS_10001

3. timestamp

Data Type: Object
Description: Date and time when the transaction occurred.
Example: 2025-01-15 14:32:10

4. home_country

Data Type: Object
Description: Country associated with the customer's registered or usual location.
Example: United Kingdom

5. source_currency

Data Type: Object
Description: Currency used by the customer to initiate the transaction.
Example: GBP

6. dest_currency

Data Type: Object
Description: Currency received at the destination of the transaction.
Example: NGN

7. channel

Data Type: Object
Description: Payment or digital channel through which the transaction was initiated.
Example: Mobile App

8. amount_src

Data Type: Object
Description: Transaction amount in the source currency.
Example: 250.00

9. amount_usd

Data Type: Float64
Description: Transaction amount converted into US dollars to provide a standardised monetary value for analysis.
Example: 315.42

10. fee

Data Type: Float64
Description: Fee charged to the customer for processing the transaction.
Example: 4.50

11. exchange_rate_src_to_dest

Data Type: Float64
Description: Exchange rate used to convert the source currency into the destination currency.
Example: 1.27

12. device_id

Data Type: Object
Description: Unique identifier for the device used to initiate the transaction.
Example: DEV_87452

13. new_device

Data Type: Boolean
Description: Indicates whether the transaction was initiated using a device that has not previously been associated with the customer.
Example: True

14. ip_address

Data Type: Object
Description: Internet Protocol (IP) address associated with the transaction.
Example: 192.168.1.25

15. ip_country

Data Type: Object
Description: Country associated with the IP address from which the transaction was initiated.
Example: Nigeria

16. location_mismatch

Data Type: Boolean
Description: Indicates whether there is a mismatch between the customer's usual location and the location detected for the transaction.
Example: True

17. ip_risk_score

Data Type: Float64
Description: Risk score associated with the IP address or IP-related transaction activity. A higher score indicates potentially greater risk.
Example: 0.82

18. kyc_tier

Data Type: Object
Description: Customer's Know Your Customer (KYC) verification level.
Example: Tier 2

19. account_age_days

Data Type: Int64
Description: Number of days the customer's account has been active at the time of the transaction.
Example: 365

20. device_trust_score

Data Type: Float64
Description: Score representing the level of trust associated with the device used for the transaction.
Example: 0.75

21. chargeback_history_count

Data Type: Int64
Description: Number of previous chargebacks associated with the customer's account.
Example: 2

22. risk_score_internal

Data Type: Float64
Description: Internal risk score generated by Finlora's existing risk assessment process for the transaction.
Example: 0.68

23. txn_velocity_1h

Data Type: Int64
Description: Number of transactions associated with the customer within the previous one-hour period.
Example: 5

24. txn_velocity_24h

Data Type: Int64
Description: Number of transactions associated with the customer within the previous 24-hour period.
Example: 18

25. corridor_risk

Data Type: Float64
Description: Risk score associated with the transaction's source and destination country or currency corridor.
Example: 0.64

26. is_fraud

Data Type: Int64
Description: Target variable indicating whether the transaction has been classified as fraudulent. A value of 1 represents a fraudulent transaction, while 0 represents a legitimate transaction.
Example: 1

In [8]:
customer_transaction_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 11400 entries, 0 to 11399
Data columns (total 26 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   transaction_id             11400 non-null  object 
 1   customer_id                11400 non-null  object 
 2   timestamp                  11371 non-null  object 
 3   home_country               11400 non-null  object 
 4   source_currency            11400 non-null  object 
 5   dest_currency              11400 non-null  object 
 6   channel                    11400 non-null  object 
 7   amount_src                 11400 non-null  object 
 8   amount_usd                 11095 non-null  float64
 9   fee                        11105 non-null  float64
 10  exchange_rate_src_to_dest  11400 non-null  float64
 11  device_id                  11400 non-null  object 
 12  new_device                 11400 non-null  bool   
 13  ip_address                 11095 non-null  obj

In [9]:
customer_transaction_data['amount_src'].value_counts(dropna=False).head(20)

amount_src
100.0     15
102.29     5
71.78      4
261.23     4
121.3      4
94.34      4
105.2      4
102.37     4
82.25      4
54.74      4
144.29     4
58.93      4
115.29     4
119.15     4
250.19     4
128.93     4
167.16     4
94.6       3
86.6       3
114.51     3
Name: count, dtype: int64

SUMMARY/OBSERVATION

1. Dataset Size: The dataset has 11,400 rows and 26 columns. All column names are formatted correctly.

2. Data Type Fixes Needed: timestamp needs to be changed from text to a date/time format. amount_sro needs to be changed from text to a number so we can do math with it.

3. Missing Data: Several columns have missing rows that we need to clean. timestamp is missing 29 values. amount_usd and ip_address are missing 305 values. fee and device_trust_score are missing 295 values. ip_country is missing 301 values, and kyc_tier is missing 300 values.

4. Data Imbalance Warning: Our target column (is_fraud) is heavily skewed. About 91% of the data is normal transactions and only 9% is fraud. If we don't fix this balance during coding, our machine learning model will just guess "not fraud" every time to get a high accuracy score, missing the actual fraud cases. We must use special training techniques to fix this.

In [10]:
#Checking for label imbalance
customer_transaction_data['is_fraud'].value_counts(normalize=True).reset_index()

,is_fraud,proportion
0,0,0.912544
1,1,0.087456


This shows that we have about 91% legitimate transactions and 9% illegitimate transactions

DATA CLEANING AND PREPARATION

In [11]:
#Converting data types and timestamp to datetime format
customer_transaction_data['timestamp'] = pd.to_datetime(customer_transaction_data['timestamp'], errors='coerce')

#Converting data type of income usd to float
customer_transaction_data['amount_src'] = pd.to_numeric(customer_transaction_data['amount_src'], errors='coerce')

In [12]:
#checking the dataset again to see if the changes have been made
customer_transaction_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 11400 entries, 0 to 11399
Data columns (total 26 columns):
 #   Column                     Non-Null Count  Dtype              
---  ------                     --------------  -----              
 0   transaction_id             11400 non-null  object             
 1   customer_id                11400 non-null  object             
 2   timestamp                  11339 non-null  datetime64[ns, UTC]
 3   home_country               11400 non-null  object             
 4   source_currency            11400 non-null  object             
 5   dest_currency              11400 non-null  object             
 6   channel                    11400 non-null  object             
 7   amount_src                 11396 non-null  float64            
 8   amount_usd                 11095 non-null  float64            
 9   fee                        11105 non-null  float64            
 10  exchange_rate_src_to_dest  11400 non-null  float64            
 11  de

Data Type Conversion: We converted timestamp to actual dates and amount_src to numbers using Pandas (errors='coerce').

Hidden Errors Found: This conversion caught corrupted or badly formatted text that wasn't noticed before. Because the computer couldn't read them, it turned them into missing values (NaN).

Increase in Missing Data: Missing rows in timestamp jumped from 29 to 61.Missing rows in amount_src went from 0 to 4.

Next Steps: These broken records were safely cleaned up in the next stage of data cleaning.

In [13]:
#Rechecking for missing values after data type conversion
customer_transaction_data.isnull().sum()

transaction_id                 0
customer_id                    0
timestamp                     61
home_country                   0
source_currency                0
dest_currency                  0
channel                        0
amount_src                     4
amount_usd                   305
fee                          295
exchange_rate_src_to_dest      0
device_id                      0
new_device                     0
ip_address                   305
ip_country                   301
location_mismatch              0
ip_risk_score                  0
kyc_tier                     300
account_age_days               0
device_trust_score           295
chargeback_history_count       0
risk_score_internal            0
txn_velocity_1h                0
txn_velocity_24h               0
corridor_risk                  0
is_fraud                       0
dtype: int64

A. Variables to Fill In (Imputation)

1. amount_usd (305 missing): We will calculate the missing values by multiplying amount_src by the currency exchange rate.

2. fee (295 missing): We will fill these using the median fee of the specific transaction channel. Any leftover missing fees will get the overall median fee.

3. ip_country (301 missing): We will fill these by copying over the user's home_country.

4. kyc_tier (300 missing): We will fill these with the most common tier (the mode).

5. device_trust_score (295 missing): We will fill these using the median score of users with the same kyc_tier and new_device status. Any leftovers will get the overall median score.

B. Variables to Delete (Dropping Rows)

1. timestamp (61 missing) & amount_src (4 missing): We will delete these rows using .dropna(). The number of missing rows is too small to matter.

2. ip_address (305 missing): We will delete these rows because IP addresses are completely unique and cannot be guessed or filled in accurately.

In [19]:
#calculating exchange rates per currency
exchange_rates = customer_transaction_data[customer_transaction_data['amount_usd'].notna()].groupby('source_currency').apply(lambda x: (x['amount_usd'] / x['amount_src'].mean()).mean()
).to_dict()
exchange_rates

/var/folders/13/c_bqr3w50896v5pg8y3p6zg80000gn/T/ipykernel_57928/2033340886.py:2: FutureWarning: DataFrameGroupBy.apply operated on the grouping columns. This behavior is deprecated, and in a future version of pandas the grouping columns will be excluded from the operation. Either pass `include_groups=False` to exclude the groupings or explicitly select the grouping columns after groupby to silence this warning.
  exchange_rates = customer_transaction_data[customer_transaction_data['amount_usd'].notna()].groupby('source_currency').apply(lambda x: (x['amount_usd'] / x['amount_src'].mean()).mean()


{'CAD': 0.7472576427098727,
 'GBP': 1.277392090620837,
 'USD': 1.0200178807388578}

In [20]:
#filling missing values in the amount_usd column using the calculated exchange rates
customer_transaction_data['amount_usd'] = customer_transaction_data.apply(lambda row: row['amount_usd'] if not pd.isna(row['amount_usd']) else row['amount_src'] * exchange_rates.get(row['source_currency'], 1), axis=1)

In [21]:
#filling the missing values of the fee variable
if 'fee' in customer_transaction_data.columns:
    customer_transaction_data['fee'] = customer_transaction_data.groupby('channel')['fee'].transform(lambda y: y.fillna(y.median()))
customer_transaction_data['fee'] = customer_transaction_data['fee'].fillna(customer_transaction_data['fee'].median())

In [24]:
#filling the missing of ip_country variable
if ('ip_country' in customer_transaction_data.columns):
    customer_transaction_data['ip_country'] = customer_transaction_data['ip_country'].fillna(customer_transaction_data['home_country'])

In [25]:
#filling the missing values of device_trust_score varaible
if 'device_trust_score' in customer_transaction_data.columns:
    if{'new_device', 'kyc_tier'}.issubset(customer_transaction_data.columns):
        customer_transaction_data['device_trust_score'] = customer_transaction_data.groupby(['new_device', 'kyc_tier'])['device_trust_score'].transform(lambda x: x.fillna(x.median()))
    customer_transaction_data['device_trust_score'] = customer_transaction_data['device_trust_score'].fillna(customer_transaction_data['device_trust_score'].median())
 

/Users/Dell/Frudulent-transaction-detection-for-Finlora-Company/finlora_env/lib/python3.13/site-packages/numpy/lib/_nanfunctions_impl.py:1213: RuntimeWarning: Mean of empty slice
  return np.nanmean(a, axis, out=out, keepdims=keepdims)
/Users/Dell/Frudulent-transaction-detection-for-Finlora-Company/finlora_env/lib/python3.13/site-packages/numpy/lib/_nanfunctions_impl.py:1213: RuntimeWarning: Mean of empty slice
  return np.nanmean(a, axis, out=out, keepdims=keepdims)


In [26]:
# dropping the timestamp, amount_src, and ip_address missing values columns as they are not needed for the analysis
customer_transaction_data.dropna(subset=['timestamp', 'amount_src', 'ip_address'], inplace=True)

In [27]:
# rechecking for missing values after data cleaning
customer_transaction_data.isnull().sum()

transaction_id               0
customer_id                  0
timestamp                    0
home_country                 0
source_currency              0
dest_currency                0
channel                      0
amount_src                   0
amount_usd                   0
fee                          0
exchange_rate_src_to_dest    0
device_id                    0
new_device                   0
ip_address                   0
ip_country                   0
location_mismatch            0
ip_risk_score                0
kyc_tier                     0
account_age_days             0
device_trust_score           0
chargeback_history_count     0
risk_score_internal          0
txn_velocity_1h              0
txn_velocity_24h             0
corridor_risk                0
is_fraud                     0
dtype: int64

STEPS TO TAKE TO DO THE SANITY CHECK

• Check for invalid numbers i.e negative numbers for variables that contains age, monetary values, velocity etc

• Check if transaction timestamp appear in the future

• Check that values falls between range

• Inspect for inconsistency among variable values in the dataset

In [28]:
#checking for negative values in numerical columns
negative_counts = {
    "amount_src": (customer_transaction_data['amount_src'] < 0).sum(),
    "amount_usd": (customer_transaction_data['amount_usd'] < 0).sum(),
    "fee": (customer_transaction_data['fee'] < 0).sum(),
    "device_trust_score": (customer_transaction_data['device_trust_score'] < 0).sum(),
    "txn_velocity_1h": (customer_transaction_data['txn_velocity_1h'] < 0).sum(),
    "txn_velocity_24h": (customer_transaction_data['txn_velocity_24h'] < 0).sum(),
    "risk_score_internal": (customer_transaction_data['risk_score_internal'] < 0).sum(),
}
negative_counts

{'amount_src': np.int64(100),
 'amount_usd': np.int64(0),
 'fee': np.int64(90),
 'device_trust_score': np.int64(190),
 'txn_velocity_1h': np.int64(190),
 'txn_velocity_24h': np.int64(0),
 'risk_score_internal': np.int64(0)}

In [30]:
#dropping rows with negative values in the specified columns
customer_transaction_data = customer_transaction_data.drop(customer_transaction_data[(customer_transaction_data['amount_src'] < 0) |
    (customer_transaction_data['amount_src'] < 0) |
    (customer_transaction_data['fee'] < 0) |
    (customer_transaction_data['device_trust_score'] < 0) |
    (customer_transaction_data['txn_velocity_1h'] < 0)].index)


In [31]:
#rechecking for negative values in numerical columns after dropping rows with negative values
negative_counts_after = {
    "amount_src": (customer_transaction_data['amount_src'] < 0).sum(),
    "amount_usd": (customer_transaction_data['amount_usd'] < 0).sum(),
    "fee": (customer_transaction_data['fee'] < 0).sum(),
    "device_trust_score": (customer_transaction_data['device_trust_score'] < 0).sum(),
    "txn_velocity_1h": (customer_transaction_data['txn_velocity_1h'] < 0).sum(),
    "txn_velocity_24h": (customer_transaction_data['txn_velocity_24h'] < 0).sum(),
    "risk_score_internal": (customer_transaction_data['risk_score_internal'] < 0).sum(),
}
negative_counts_after

{'amount_src': np.int64(0),
 'amount_usd': np.int64(0),
 'fee': np.int64(0),
 'device_trust_score': np.int64(0),
 'txn_velocity_1h': np.int64(0),
 'txn_velocity_24h': np.int64(0),
 'risk_score_internal': np.int64(0)}

In [32]:
#checking for future timestamps in the dataset
customer_transaction_data[customer_transaction_data['timestamp'] > pd.Timestamp.utcnow()]

,transaction_id,customer_id,timestamp,home_country,source_currency,dest_currency,channel,amount_src,amount_usd,fee,...,ip_risk_score,kyc_tier,account_age_days,device_trust_score,chargeback_history_count,risk_score_internal,txn_velocity_1h,txn_velocity_24h,corridor_risk,is_fraud


In [33]:
#checking for location mismatch
customer_transaction_data['location_mismatch'].value_counts().reset_index()

,location_mismatch,count
0,False,9047
1,True,1793


In [34]:
#reviewing the channel variables
customer_transaction_data['channel'].unique()

array(['ATM', 'web', 'mobile', 'WEB', ' web  ', 'MOBILE', 'unknown',
       'mobille', ' mobile  ', 'weeb', 'ATm', ' ATM  '], dtype=object)

In [35]:
#formatting the channel variable to have consistent values
customer_transaction_data['channel'] = customer_transaction_data['channel'].str.lower().str.strip()
customer_transaction_data['channel'].unique()

array(['atm', 'web', 'mobile', 'unknown', 'mobille', 'weeb'], dtype=object)

In [39]:
#Replacing incorrect spelling of 'mobile' and 'web' in the channel variable
customer_transaction_data['channel'] = customer_transaction_data['channel'].replace({
    'mobille': 'mobile', 
    'weeb': 'web'
    })

In [50]:
#replacing unkown values in the channel variable with NaN
customer_transaction_data['channel'] = customer_transaction_data['channel'].replace({'unknown': np.nan})
customer_transaction_data['channel'].unique()

array(['atm', 'web', 'mobile', nan], dtype=object)

In [42]:
#reviewing the source_currency variable
customer_transaction_data['source_currency'].unique()

array(['USD', 'CAD', 'GBP'], dtype=object)

In [44]:
#reviewing the destination_currency variable
customer_transaction_data['dest_currency'].unique()

array(['CAD', 'MXN', 'CNY', 'EUR', 'INR', 'GBP', 'PHP', 'NGN', 'USD'],
      dtype=object)

In [45]:
#reviewing the kyc_tier variable
customer_transaction_data['kyc_tier'].unique()

array(['standard', 'enhanced', 'low', ' standard  ', 'standrd',
       ' enhanced  ', 'STANDARD', 'unknown', 'enhancd', ' low  ',
       'ENHANCED', 'LOW'], dtype=object)

In [46]:
#formatting the kyc_tier variable to have consistent values
customer_transaction_data['kyc_tier'] = customer_transaction_data['kyc_tier'].str.lower().str.strip()
customer_transaction_data['kyc_tier'].unique()

array(['standard', 'enhanced', 'low', 'standrd', 'unknown', 'enhancd'],
      dtype=object)

In [47]:
#Replacing incorrect spelling of 'standard' and 'enhanced' in the channel variable
customer_transaction_data['kyc_tier'] = customer_transaction_data['kyc_tier'].replace({
    'standrd': 'standard', 
    'enhancd': 'enhanced'
    })

In [49]:
#replacing unkown values in the kyc_tier variable with NaN
customer_transaction_data['kyc_tier'] = customer_transaction_data['kyc_tier'].replace({'unknown': np.nan})
customer_transaction_data['kyc_tier'].unique()

array(['standard', 'enhanced', 'low', nan], dtype=object)

In [51]:
customer_transaction_data.isnull().sum()

transaction_id                0
customer_id                   0
timestamp                     0
home_country                  0
source_currency               0
dest_currency                 0
channel                      36
amount_src                    0
amount_usd                    0
fee                           0
exchange_rate_src_to_dest     0
device_id                     0
new_device                    0
ip_address                    0
ip_country                    0
location_mismatch             0
ip_risk_score                 0
kyc_tier                     29
account_age_days              0
device_trust_score            0
chargeback_history_count      0
risk_score_internal           0
txn_velocity_1h               0
txn_velocity_24h              0
corridor_risk                 0
is_fraud                      0
dtype: int64

In [52]:
customer_transaction_data.dropna(inplace=True)
customer_transaction_data.info()

<class 'pandas.core.frame.DataFrame'>
Index: 10780 entries, 0 to 11399
Data columns (total 26 columns):
 #   Column                     Non-Null Count  Dtype              
---  ------                     --------------  -----              
 0   transaction_id             10780 non-null  object             
 1   customer_id                10780 non-null  object             
 2   timestamp                  10780 non-null  datetime64[ns, UTC]
 3   home_country               10780 non-null  object             
 4   source_currency            10780 non-null  object             
 5   dest_currency              10780 non-null  object             
 6   channel                    10780 non-null  object             
 7   amount_src                 10780 non-null  float64            
 8   amount_usd                 10780 non-null  float64            
 9   fee                        10780 non-null  float64            
 10  exchange_rate_src_to_dest  10780 non-null  float64            
 11  device_

In [53]:
#reviewing the home country variable
customer_transaction_data['home_country'].unique()

array(['US', 'CA', 'UK', ' UK  ', ' US  ', 'unknown', ' CA  '],
      dtype=object)

In [54]:
#Formatting the home_country variable to have consistent values
customer_transaction_data['home_country'] = customer_transaction_data['home_country'].str.upper().str.strip()
customer_transaction_data['home_country'].unique()

array(['US', 'CA', 'UK', 'UNKNOWN'], dtype=object)

In [60]:
#replacing unkown values in the home_country variable with NaN
customer_transaction_data['home_country'] = customer_transaction_data['home_country'].replace({'UNKNOWN': np.nan})
customer_transaction_data['home_country'].unique()

array(['US', 'CA', 'UK', nan], dtype=object)

In [56]:
#reviwing the ip_country variable
customer_transaction_data['ip_country'].unique()

array(['US', 'CA', 'UK', ' US  ', 'unknown', ' CA  ', ' UK  '],
      dtype=object)

In [57]:
#formatting the ip_country variable to have consistent values
customer_transaction_data['ip_country'] = customer_transaction_data['ip_country'].str.upper().str.strip()
customer_transaction_data['ip_country'].unique()

array(['US', 'CA', 'UK', 'UNKNOWN'], dtype=object)

In [58]:
#replacing unkown values in the ip_country variable with NaN
customer_transaction_data['ip_country'] = customer_transaction_data['ip_country'].replace({'UNKNOWN': np.nan})
customer_transaction_data['ip_country'].unique()

array(['US', 'CA', 'UK', nan], dtype=object)

In [65]:
#Create a directory to store artifacts in Finlora_Dataset Folder
import os
os.makedirs("../Finlora_Dataset/Artifacts", exist_ok=True)

In [68]:
customer_transaction_data.to_csv(r'../Finlora_Dataset/Artifacts/cleaned_cleaned_data.csv')